crates/storage/src/transactional.rs

```diff
@@ -1,96 +1,478 @@
+            .insert(k, WriteOperation::Insert(Arc::new(buf.to_vec())));
+        Ok(buf.len())
+    }
+
+    fn take(&mut self, key: &[u8], column: Self::Column) -> StorageResult<Option<Value>> {
+        let k = key.to_vec();
+        let entry = self.changes.entry(column.id()).or_default().entry(k);
+
+        match entry {
+            Entry::Occupied(mut occupied) => {
+                let old = occupied.insert(WriteOperation::Remove);
+
+                match old {
+                    WriteOperation::Insert(value) => Ok(Some(value)),
+                    WriteOperation::Remove => Ok(None),
+                }
+            }
+            Entry::Vacant(vacant) => {
+                vacant.insert(WriteOperation::Remove);
+                self.storage.get(key, column)
+            }
+        }
+    }
+
+    fn delete(&mut self, key: &[u8], column: Self::Column) -> StorageResult<()> {
+        let k = key.to_vec();
+        self.changes
+            .entry(column.id())
+            .or_default()
+            .insert(k, WriteOperation::Remove);
+        Ok(())
     }
 }
 
-impl<Storage: ?Sized> AsMut<Storage> for StorageTransaction<Storage> {
-    fn as_mut(&mut self) -> &mut Storage {
-        (*self.transaction).as_mut()
+impl<Column, S> BatchOperations for StorageTransactionInner<S>
+where
+    Column: StorageColumn,
+    S: KeyValueInspect<Column = Column>,
+{
+    fn batch_write<I>(&mut self, column: Column, entries: I) -> StorageResult<()>
+    where
+        I: Iterator<Item = (Vec<u8>, WriteOperation)>,
+    {
+        let btree = self.changes.entry(column.id()).or_default();
+        entries.for_each(|(key, operation)| {
+            btree.insert(key, operation);
+        });
+        Ok(())
     }
 }
 
-impl<Storage: ?Sized> StorageTransaction<Storage> {
-    /// Committing of the state consumes `Self`.
-    pub fn commit(mut self) -> StorageResult<()> {
-        self.transaction.commit()
+// The `IterableStore` should only be implemented for the actual storage,
+// not the storage transaction, to maximize the performance.
+//
+// Another reason is that only actual storage with finalized updates should be
+// used to iterate over its entries to avoid inconsistent results.
+//
+// We implement `IterableStore` for `StorageTransactionInner` only to allow
+// using it in the tests and benchmarks as a type(not even implementation of it).
+#[cfg(feature = "test-helpers")]
+impl<S> IterableStore for StorageTransactionInner<S>
+where
+    S: IterableStore,
+{
+    fn iter_store(
+        &self,
+        _: Self::Column,
+        _: Option<&[u8]>,
+        _: Option<&[u8]>,
+        _: IterDirection,
+    ) -> BoxedIter<KVItem> {
+        unimplemented!()
     }
 }
 
-/// Provides a view of the storage at the given height.
-/// It guarantees to be atomic, meaning the view is immutable to outside modifications.
-pub trait AtomicView: Send + Sync {
-    /// The type of the storage view.
-    type View;
+#[cfg(feature = "test-helpers")]
+mod test {
```

Comment:
I think all these tests could us a little clarification on what's being tested and what the expected results are. They're pretty close.